src/sage/rings/species.py

```diff
@@ -0,0 +1,1268 @@
+        ElementCache.__init__(self)
+        self._k = indices._k
+
+    def _project(self, G, pi, part):
+        r"""
+        Project `G` onto a subset ``part`` of its domain.
+
+        ``part`` must be a union of cycles, but this is not checked.
+        """
+        restricted_gens = [[cyc for cyc in gen.cycle_tuples() if cyc[0] in part] for gen in G.gens()]
+        mapping = dict()
+        for k, v in pi.items():
+            es = [e for e in v if e in part]
+            if es:
+                mapping[k] = es
+        return PermutationGroup(gens=restricted_gens, domain=part), mapping
+
+    def _element_constructor_(self, G, pi=None):
+        r"""
+        Construct the `k`-variate molecular species with the given data.
+
+        INPUT:
+
+        - ``G`` - an element of ``self`` (in this case pi must be ``None``)
+          or a permutation group, or a pair ``(X, a)`` consisting of a
+          finite set and an action.
+        - ``pi`` - a dict mapping sorts to iterables whose union is the
+          domain of ``G`` (if ``G`` is a permutation group) or `X` (if ``G``)
+          is a pair ``(X, a)``. If `k=1`, `pi` can be omitted.
+
+        If `G = (X, a)`, then `X` should be a finite set and `a` a transitive
+        action of `G` on `X`.
+
+        EXAMPLES::
+
+            sage: P = PolynomialSpecies(ZZ, ["X", "Y"])
+            sage: P(SymmetricGroup(4).young_subgroup([2, 2]), {1: [1,2], 2: [3,4]})
+            E_2(X)*E_2(Y)
+
+            sage: X = SetPartitions(4, 2)
+            sage: a = lambda g, x: SetPartition([[g(e) for e in b] for b in x])
+            sage: P((X, a), {1: [1,2], 2: [3,4]})
+            X^2*E_2(Y) + X^2*Y^2 + E_2(X)*Y^2 + E_2(X)*E_2(Y)
+
+        TESTS::
+
+            sage: P = PolynomialSpecies(ZZ, ["X", "Y"])
+            sage: M = P._indices
+            sage: M(CyclicPermutationGroup(4), {1: [1,2], 2: [3,4]})
+            Traceback (most recent call last):
+            ...
+            ValueError: For each orbit of Cyclic group of order 4 as a permutation group, all elements must belong to the same sort
+        """
+        if parent(G) == self:
+            if pi is not None:
+                raise ValueError("cannot reassign sorts to a molecular species")
+            return G
+        if isinstance(G, PermutationGroup_generic):
+            if pi is None:
+                if self._k == 1:
+                    pi = {1: G.domain()}
+                else:
+                    raise ValueError("the assignment of sorts to the domain elements must be provided")
+            if not set(pi.keys()).issubset(range(1, self._k + 1)):
+                raise ValueError(f"keys of pi must be in the range [1, {self._k}]")
+            if sum(len(p) for p in pi.values()) != len(G.domain()) or set(chain.from_iterable(pi.values())) != set(G.domain()):
+                raise ValueError("values of pi must partition X")
```

Comment:
This message (and similar messages) should be improved.  In particular, it might be unclear what `pi` and `X` are and should be.